File: geometry.go
package main

import "math"

func newXZSquare(size float64, shader Shader) *Model {
	v0 := Vector3{x: -size / 2, y: 0, z: -size / 2}
	v1 := Vector3{x: size / 2, y: 0, z: -size / 2}
	v2 := Vector3{x: size / 2, y: 0, z: size / 2}
	v3 := Vector3{x: -size / 2, y: 0, z: size / 2}

	t0 := []Vector3{v1, v0, v2}
	t1 := []Vector3{v3, v2, v0}

	normal := Vector3{x: 0, y: 1, z: 0}

	textt0v0 := []float64{0, 0, 0}
	textt1v0 := []float64{0, 0, 0}
	textt0v1 := []float64{0.999, 0, 0}
	textt0v2 := []float64{0.999, 0.999, 0}
	textt1v2 := []float64{0.999, 0.999, 0}
	textt1v3 := []float64{0, 0.999, 0}

	triangle1 := newTriangle()
	triangle1.worldVerts = t0
	triangle1.normals = []Vector3{normal, normal, normal}
	triangle1.uvMapping = [][]float64{textt0v1, textt0v0, textt0v2}

	triangle2 := newTriangle()
	triangle2.worldVerts = t1
	triangle2.normals = []Vector3{normal, normal, normal}
	triangle2.uvMapping = [][]float64{textt1v3, textt1v2, textt1v0}

	return &Model{
		triangles: []*Triangle{triangle1, triangle2},
		shader:    shader,
	}
}

func newXYSquare(size float64, shader Shader) *Model {
	pos := size / 2
	neg := -size / 2

	v0 := Vector3{x: neg, y: pos, z: 0}
	v1 := Vector3{x: neg, y: neg, z: 0}
	v2 := Vector3{x: pos, y: neg, z: 0}
	v3 := Vector3{x: pos, y: pos, z: 0}

	t0 := []Vector3{v1, v2, v0}
	t1 := []Vector3{v3, v0, v2}

	normal := Vector3{x: 0, y: 0, z: 1}

	textt0v0 := []float64{0, 0.999, 0}
	textt1v0 := []float64{0, 0.999, 0}
	textt0v1 := []float64{0, 0, 0}
	textt0v2 := []float64{0.999, 0, 0}
	textt1v2 := []float64{0.999, 0, 0}
	textt1v3 := []float64{0.999, 0.999, 0}

	triangle1 := newTriangle()
	triangle1.worldVerts = t0
	triangle1.normals = []Vector3{normal, normal, normal}
	triangle1.uvMapping = [][]float64{textt0v1, textt0v2, textt0v0}

	triangle2 := newTriangle()
	triangle2.worldVerts = t1
	triangle2.normals = []Vector3{normal, normal, normal}
	triangle2.uvMapping = [][]float64{textt1v3, textt1v0, textt1v2}

	return &Model{
		triangles: []*Triangle{triangle1, triangle2},
		shader:    shader,
	}
}

func newYZSquare(size float64, shader Shader) *Model {
	pos := size / 2
	neg := -size / 2

	v0 := Vector3{x: 0, y: pos, z: neg}
	v1 := Vector3{x: 0, y: neg, z: neg}
	v2 := Vector3{x: 0, y: neg, z: pos}
	v3 := Vector3{x: 0, y: pos, z: pos}

	t0 := []Vector3{v1, v2, v0}
	t1 := []Vector3{v3, v0, v2}

	normal := Vector3{x: 0, y: 0, z: 1}

	textt0v0 := []float64{0, 0.999, 0}
	textt1v0 := []float64{0, 0.999, 0}
	textt0v1 := []float64{0, 0, 0}
	textt0v2 := []float64{0.999, 0, 0}
	textt1v2 := []float64{0.999, 0, 0}
	textt1v3 := []float64{0.999, 0.999, 0}

	triangle1 := newTriangle()
	triangle1.worldVerts = t0
	triangle1.normals = []Vector3{normal, normal, normal}
	triangle1.uvMapping = [][]float64{textt0v1, textt0v2, textt0v0}

	triangle2 := newTriangle()
	triangle2.worldVerts = t1
	triangle2.normals = []Vector3{normal, normal, normal}
	triangle2.uvMapping = [][]float64{textt1v3, textt1v0, textt1v2}

	return &Model{
		triangles: []*Triangle{triangle1, triangle2},
		shader:    shader,
	}
}

func newCube(size float64, shader Shader) *Model {
	triangles := []*Triangle{}

	triangles = append(triangles, newXZSquare(size, shader).rotateX(math.Pi).moveY(-size/2).triangles...) // bottom
	triangles = append(triangles, newXZSquare(size, shader).moveY(size/2).triangles...)                   // top

	triangles = append(triangles, newYZSquare(size, shader).rotateY(math.Pi).moveX(size/2).triangles...) // right
	triangles = append(triangles, newYZSquare(size, shader).moveX(-size/2).triangles...)                 // left

	triangles = append(triangles, newXYSquare(size, shader).rotateY(math.Pi).moveZ(-size/2).triangles...) // back
	triangles = append(triangles, newXYSquare(size, shader).moveZ(size/2).triangles...)                   // front

	return &Model{
		triangles: triangles,
		shader:    shader,
	}
}

File: src/geometry.rs
use crate::algebra::*;
use crate::model::*;
use crate::shader::*;
use std::f32::consts::PI;
use std::rc::Rc;

pub fn new_xz_square(size: f32, shader: Rc<dyn Shader>) -> Model {
    let v0 = vec3f(-size / 2., 0., -size / 2.);
    let v1 = vec3f(size / 2., 0., -size / 2.);
    let v2 = vec3f(size / 2., 0., size / 2.);
    let v3 = vec3f(-size / 2., 0., size / 2.);

    let t0 = [v1, v0, v2];
    let t1 = [v3, v2, v0];

    let normal = vec3f(0., 1., 0.);

    let textt0v0 = [0., 0., 0.];
    let textt1v0 = [0., 0., 0.];
    let textt0v1 = [0.999, 0., 0.];
    let textt0v2 = [0.999, 0.999, 0.];
    let textt1v2 = [0.999, 0.999, 0.];
    let textt1v3 = [0., 0.999, 0.];

    let mut triangle1 = Triangle::new();
    triangle1.world_verts = t0;
    triangle1.normals = [normal, normal, normal];
    triangle1.uv_mapping = [textt0v1, textt0v0, textt0v2];

    let mut triangle2 = Triangle::new();
    triangle2.world_verts = t1;
    triangle2.normals = [normal, normal, normal];
    triangle2.uv_mapping = [textt1v3, textt1v2, textt1v0];

    return Model::new(vec![triangle1, triangle2], shader);
}

pub fn new_xy_square(size: f32, shader: Rc<dyn Shader>) -> Model {
    let pos = size / 2.;
    let neg = -size / 2.;

    let v0 = Vec3f { x: neg, y: pos, z: 0. };
    let v1 = Vec3f { x: neg, y: neg, z: 0. };
    let v2 = Vec3f { x: pos, y: neg, z: 0. };
    let v3 = Vec3f { x: pos, y: pos, z: 0. };

    let t0 = [v1, v2, v0];
    let t1 = [v3, v0, v2];

    let normal = Vec3f { x: 0., y: 0., z: 1. };

    let textt0v0 = [0., 0.999, 0.];
    let textt1v0 = [0., 0.999, 0.];
    let textt0v1 = [0., 0., 0.];
    let textt0v2 = [0.999, 0., 0.];
    let textt1v2 = [0.999, 0., 0.];
    let textt1v3 = [0.999, 0.999, 0.];

    let mut triangle1 = Triangle::new();
    triangle1.world_verts = t0;
    triangle1.normals = [normal, normal, normal];
    triangle1.uv_mapping = [textt0v1, textt0v2, textt0v0];

    let mut triangle2 = Triangle::new();
    triangle2.world_verts = t1;
    triangle2.normals = [normal, normal, normal];
    triangle2.uv_mapping = [textt1v3, textt1v0, textt1v2];

    return Model::new(vec![triangle1, triangle2], shader);
}

pub fn new_yz_square(size: f32, shader: Rc<dyn Shader>) -> Model {
    let pos = size / 2.;
    let neg = -size / 2.;

    let v0 = Vec3f { x: 0., y: pos, z: neg };
    let v1 = Vec3f { x: 0., y: neg, z: neg };
    let v2 = Vec3f { x: 0., y: neg, z: pos };
    let v3 = Vec3f { x: 0., y: pos, z: pos };

    let t0 = [v1, v2, v0];
    let t1 = [v3, v0, v2];

    let normal = Vec3f { x: 0., y: 0., z: 1. };

    let textt0v0 = [0., 0.999, 0.];
    let textt1v0 = [0., 0.999, 0.];
    let textt0v1 = [0., 0., 0.];
    let textt0v2 = [0.999, 0., 0.];
    let textt1v2 = [0.999, 0., 0.];
    let textt1v3 = [0.999, 0.999, 0.];

    let mut triangle1 = Triangle::new();
    triangle1.world_verts = t0;
    triangle1.normals = [normal, normal, normal];
    triangle1.uv_mapping = [textt0v1, textt0v2, textt0v0];

    let mut triangle2 = Triangle::new();
    triangle2.world_verts = t1;
    triangle2.normals = [normal, normal, normal];
    triangle2.uv_mapping = [textt1v3, textt1v0, textt1v2];

    return Model::new(vec![triangle1, triangle2], shader);
}

pub fn new_cube(size: f32, shader: Rc<dyn Shader>) -> Model {
    let mut bottom = new_xz_square(size, Rc::clone(&shader));
    bottom.rotate_x(PI).move_y(-size / 2.);

    let mut up = new_xz_square(size, Rc::clone(&shader));
    up.move_y(size / 2.);

    let mut right = new_yz_square(size, Rc::clone(&shader));
    right.rotate_y(PI).move_x(size / 2.);

    let mut left = new_yz_square(size, Rc::clone(&shader));
    left.move_x(-size / 2.);

    let mut back = new_xy_square(size, Rc::clone(&shader));
    back.rotate_y(PI).move_z(-size / 2.);

    let mut front = new_xy_square(size, Rc::clone(&shader));
    front.move_z(size / 2.);

    let mut triangles = vec![];
    triangles.append(&mut bottom.triangles);
    triangles.append(&mut up.triangles);
    triangles.append(&mut left.triangles);
    triangles.append(&mut right.triangles);
    triangles.append(&mut front.triangles);
    triangles.append(&mut back.triangles);

    return Model::new(triangles, shader);
}
